proof/crefine/AARCH64/Arch_C.thy

```diff
@@ -4405,6 +4406,300 @@ lemma decodeSGISignalInvocation_ccorres:
+      apply (rule ccorres_from_vcg_throws[where P=\<top> and P'=UNIV])
+      apply (rule allI, rule conseqPre, vcg)
+      apply (clarsimp simp: fst_throwError_returnOk exception_defs syscall_error_rel_def
+                            syscall_error_to_H_cases)
+     apply (prop_tac "\<not>(smc_badge \<noteq> 0 \<and> smc_badge \<noteq> args ! 0)")
+      apply (solves \<open>clarsimp simp: ccap_relation_SMCCap_lift_eq\<close>)
+     apply (simp only:) (* Avoid rewriting condition above, so whenE can rewrite to False *)
+     apply (clarsimp simp: whileAnno_def)
+     apply csymbr
+     apply (while_unroll_syscall 0)
+       apply (while_unroll_syscall 1)
+         apply (while_unroll_syscall 2)
+           apply (while_unroll_syscall 3)
+             apply (while_unroll_syscall 4)
+               apply (while_unroll_syscall 5)
+                 apply (while_unroll_syscall 6)
+                   apply (while_unroll_syscall 7)
+                     apply (rule ccorres_basic_srnoop2, simp)
+                     apply (rule ccorres_expand_while_iff_Seq[THEN iffD1])
+                     apply (rule ccorres_cond_false)
+                     apply (clarsimp simp: returnOk_def ccorres_invocationCatch_Inr)
+                     apply (ctac (no_vcg) add: setThreadState_ccorres)
+                      apply (clarsimp simp: performInvocation_def AARCH64_H.performInvocation_def
+                                            liftE_bindE performSMCInvocation_always_replies)
+                      apply (rule ccorres_add_return2)
+                      apply (ctac (no_vcg) add: invokeSMCCall_ccorres)
+                       apply (rule ccorres_from_vcg_throws[where P=\<top> and P'=UNIV])
+                       apply (rule allI, rule conseqPre, vcg)
+                       apply (clarsimp simp: return_def)
+                      apply wp
+                     apply (wpsimp wp: sts_invs_minor' ct_in_state'_set)
+                   apply (wp,
+                          clarsimp, rule conseqPre,
+                          vcg exspec=getSyscallArg_modifies, rule subset_refl)+
+    apply wp
+    apply (clarsimp cong: conj_cong)
+    apply assumption
+   apply (vcg exspec=getSyscallArg_modifies)
+  apply (clarsimp simp: rf_sr_ksCurThread cap_get_tag_isCap_unfolded_H_cap2)
+  apply (rule conjI; clarsimp)
+   apply (prop_tac "8 \<le> length args", solves \<open>clarsimp simp: not_less word_le_nat_alt\<close>)
+   apply (prop_tac "args \<noteq> []", fastforce)
+   apply (clarsimp simp: sysargs_rel_to_n ct_active_st_tcb_at_minor')
+   apply fastforce
+  (* 8 times array update = smc_args_to_C *)
+  apply (rule drop_imp, rule allI)
+  apply (rename_tac smc_arg, case_tac smc_arg)
+  apply (clarsimp simp: smc_args_to_C_def intro!: array_ext)
+  apply (rename_tac i, case_tac i, simp)
+  apply (repeat 7 \<open>rename_tac i', case_tac i', simp add: eval_nat_numeral\<close>)
+  apply simp
+  done
+
+end (* of private method context *)
```

Comment:
yes, like this! (related to my other comment about a method context, forgot where it was again)